poly1305/src/avx2.rs

```diff
@@ -0,0 +1,157 @@
+//! AVX2 implementation of the Poly1305 state machine.
+
+// The State struct and its logic was originally derived from Goll and Gueron's AVX2 C
+// code:
+//     [Vectorization of Poly1305 message authentication code](https://ieeexplore.ieee.org/document/7113463)
+//
+// which was sourced from Bhattacharyya and Sarkar's modified variant:
+//     [Improved SIMD Implementation of Poly1305](https://eprint.iacr.org/2019/842)
+//     https://github.com/Sreyosi/Improved-SIMD-Implementation-of-Poly1305
+//
+// The logic has been extensively rewritten and documented, and several bugs in the
+// original C code were fixed.
+//
+// Note that State only implements the original Goll-Gueron algorithm, not the
+// optimisations provided by Bhattacharyya and Sarkar. The latter require the message
+// length to be known, which is incompatible with the streaming API of UniversalHash.
+
+use universal_hash::generic_array::GenericArray;
+
+use crate::{Block, Key, Tag, BLOCK_SIZE};
+
+mod helpers;
+use self::helpers::*;
+
+const BLOCK_X4_SIZE: usize = BLOCK_SIZE * 4;
+
+#[derive(Clone)]
+struct Initialized {
+    p: Aligned4x130,
+    m: SpacedMultiplier4x130,
+    r4: PrecomputedMultiplier,
+}
+
+#[derive(Clone)]
+pub(crate) struct State {
+    k: AdditionKey,
+    r1: PrecomputedMultiplier,
+    r2: PrecomputedMultiplier,
+    initialized: Option<Initialized>,
+    cached_blocks: [u8; BLOCK_X4_SIZE],
+    num_cached_blocks: usize,
+    partial_block: Option<Block>,
+}
+
+impl State {
+    /// Initialize Poly1305 state with the given key
+    pub(crate) fn new(key: &Key) -> Self {
+        // Prepare addition key and polynomial key.
+        let (k, r1) = prepare_keys(key);
+
+        // Precompute R^2.
+        let r2 = (r1 * r1).reduce();
+
+        State {
+            k,
+            r1,
+            r2: r2.into(),
+            initialized: None,
+            cached_blocks: [0u8; BLOCK_X4_SIZE],
+            num_cached_blocks: 0,
+            partial_block: None,
+        }
+    }
+
+    /// Reset internal state
+    pub(crate) fn reset(&mut self) {
+        self.initialized = None;
+        self.num_cached_blocks = 0;
+    }
+
+    pub(crate) fn compute_block(&mut self, block: &Block, partial: bool) {
+        // We can cache a single partial block.
+        if partial {
+            assert!(self.partial_block.is_none());
+            self.partial_block = Some(*block);
+            return;
+        }
+
+        self.cached_blocks
+            [self.num_cached_blocks * BLOCK_SIZE..(self.num_cached_blocks + 1) * BLOCK_SIZE]
+            .copy_from_slice(block);
```

Comment:
This seems like the sort of thing that might benefit from the cache being a `[Block; 4]`, although it seems it might complicate some of the cases below. Six of one, half dozen of the other.

Reply:
I went for a single `[u8; 4 * BLOCK_SIZE]` because `_mm256_loadu_si256` (inside `Aligned4x130::from_blocks`) takes a pointer to a contiguous 32-byte array, and I wasn't sure whether `[Block; 4]` would be contiguous.

Reply:
To my knowledge, Rust array layout is fully guaranteed in these cases (i.e. size = stride), even with `generic-array` (which is just a wrapper for the underlying array types which uses type-level operations to choose which size to use).

I believe safe conversions for things like `&[[T; N]] -> &[T]` are one of the goals of the [safe transmute project](https://internals.rust-lang.org/t/pre-rfc-v2-safe-transmute/11431).

Reply:
More info here: https://github.com/rust-lang/unsafe-code-guidelines/blob/1c44b54/reference/src/layout/arrays-and-slices.md

Reply:
I'll go ahead and merge. I can take a look at following up on this before cutting a release unless you'd like to.